### ISSUE TITLE
Neural Net Helpers

### ISSUE THREAD

**Issue by null-a** [probmods/webppl#710]
(I'm splitting this out from #663.)

Idea: by leveraging our parameter/guide infrastructure we can use function composition to build simple neural networks in webppl. This would look something like this:

```js
// make a net:
var net = compose(sigmoid, linear(10, 'name'));
// use it:
net(ones([5,1]));
// slightly fuller example:
var predictNet = stack([linear(10, 'net1'), tanh, linear(2, 'net2'), sigmoid]);
var guideParams = {ps: predict(x)};
// etc.
```

Under the hood, the parameter infrastructure takes care of doing the right thing with parameters, and we determine whether we need to put a prior over parameters by checking whether we're in a guide thunk. (By building on #652.) ~~There's a work-in-progress implementation of this [here](https://github.com/probmods/webppl/compare/dev...null-a:wppl-nets).~~

For the user, the main advantage of this over using adnn nets is that there's no need to use `nneval` or call `setTraining` on the net. Also, with this approach, it's possible to mimic the common pattern of `cache`ing a fragment of network by simply reusing network names where sharing is required.

One question I'd like to answer (in part to help convince myself that this is flexible enough to handle future requirements) is how we best handle weight decay with this scheme. You can imagine specifying weight decay with something like `linear({..., weightDecay: 'l2'})`, and for nets that are part of the model we would implement this by switching from using `ImproperUniform` to `TensorGaussian` as the prior. But for nets that are part of the guide we don't have this option, so we need to do something else. (@dritchie and I have [discussed](https://gist.github.com/null-a/a42a855efbce27da1f8c4f4248ddd790) this a little already.)

**Comment by dritchie** [comment]
Re: weight decay for for guide nets: can we not just use the 'prior swapping' trick? I realize that this would mean that we'd be using a "guide inside of a guide," but there's no reason this wouldn't work, right? Currently (i.e. we haven't switched to using guide thunks yet), there's no notion of being 'inside' a guide, so something like this would work:
```js
var data = // some tensor

var guideParamWithPrior = function(data) {
   var W = sample(TensorGaussian({ mu: 0, sigma: 1, dims: [1, data.length] }), {
      guide: Delta({ v: param({ name: 'W', dims: [1, data.length]) })
   });
   var b = sample(Gaussian({mu: 0, sigma: 1}), {
      guide: Delta({ v: param({ name: 'b' })
   };
   return T.sigmoid(T.get(T.dot(W, data), 0) + b);
};

var x = sample(Bernoulli({ p: 0.5 }), {
   guide: Bernoulli({ p: guideParamWithPrior(data) } )
});
```
(In practice, the stuff in `guideParamWithPrior` would be hidden behind one of the net creation functions). Is there a compelling reason to forbid this sort of pattern once we switch to using guide thunks? (e.g. is there some way this makes the estimator math subtly wrong that I'm not seeing?)

**Comment by null-a** [comment]
In case anyone happens to be following along, my work-in-progress version of this is now in the [webppl-nn](https://github.com/null-a/webppl-nn) package. This needs #715 merging and some more work before it could be generally useful. The code has quite a lot  of comments in it though, should anyone be interested in the general idea.

**Comment by null-a** [comment]
I now have what I think is a workable implementation of weight decay (for both model and guide parameters) in `webppl-nn`. It's built on top of a [small change to webppl](https://github.com/null-a/webppl/commit/78409804a9393fde594bf4a71eabbc8b718aa392) that allows extra terms to be added to the optimization objective from within a webppl program.

The code is on [this branch](https://github.com/null-a/webppl-nn/tree/weight-decay). Here's a [description](https://github.com/null-a/webppl-nn/tree/weight-decay#regparaml2sd) of the new method used to create regularized params.

My plan is to test this locally a little more, and then open a PR for the webppl change in a few days time.
